proposals/p2173.md

```diff
@@ -0,0 +1,1291 @@
+        interface. Under this proposal, the interface of a type is not affected
+        by coincidental equality, only by intentional assignment to an
+        associated type.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Using `=` rather than `==` in `impl`s is easier to read and write.
+    -   The interface available on a type is more predictable in this proposal
+        than in the status quo ante, making code easier to understand.
+
+## Alternatives considered
+
+### Status quo
+
+The [Problem](#problem) section describes some concerns with the status quo
+approach. This proposal aims to address those concerns.
+
+### Restrict constraints to allow computable type equality
+
+We could somehow restrict which constraints can be specified, in order to ensure
+that our type equality rule is efficiently computable, perhaps even in a way
+that permits a deterministic computation of a canonical type. The exact details
+of how we would do this have not been worked out, but this might lead to a
+coherent rule that has transitive type equality. However, this would not solve
+the "equal types with different interfaces" problem.
+
+### Find a fully transitive approach to type equality
+
+This proposal makes type equality transitive, but still has non-transitive
+"same-type" constraints describing types that are known to always eventual be
+the same after substitution, but which are not treated as the same type value
+while type-checking a generic. This proposal also imposes a directionality on
+rewrites, a restriction to only rewrite at a depth of exactly one associated
+constant, and a restriction that only one value can be specified as equal to a
+given associated constant. We could seek a type equality rule that would avoid
+having two different kinds of equality and would avoid some or all of the
+restrictions placed on rewrite constraints.
+
+One candidate approach is to accept that the full generality of the equality
+problem is
+[hard](https://link.springer.com/content/pdf/10.1007/3-540-17220-3_7.pdf), but
+attempt to solve it anyway. Because of the close correspondence between Turing
+machines and string rewriting systems, this results in a type system with no
+upper bound on its runtime. We consider this unacceptable for Carbon, but it is
+the approach taken by some other modern languages:
+
+-   Swift has an
+    [undecidable type system](https://forums.swift.org/t/swift-type-checking-is-undecidable/39024)
+    due to this, and handles the problem by placing a limit on the complexity of
+    cases they will accept, but that seems unsuited to Carbon's goals, as the
+    rule cannot be readily understood by a developer nor effectively worked
+    around.
+-   Rust also has an
+    [undecidable type system (content warning: swear words)](https://sdleffler.github.io/RustTypeSystemTuringComplete/)
```

Comment:
Trivial nit: slightly more precise / clear CW?

```suggestion
    [undecidable type system (content warning: profanity)](https://sdleffler.github.io/RustTypeSystemTuringComplete/)
```

Also, wow, this is such a rabbit hole of esoteric languages...

Reply:
I'd like to push back on this suggestion just a little: in this context, the word in question is not being used as profanity but simply as part of a proper noun, but it is nonetheless objectively a swear word. Happy to accept the suggestion if you still think it's a beneficial change given that argument.

Reply:
I see your point there.

I struggle with "swear words" because for me it seems like a slang term or idiom and doesn't unambiguously communicate the severity of the word choice. A result is that I'm not sure it'll serve its purpose reliably.

Maybe "obscenities" or "obscene words" or "obscene language" would work to avoid talking about the *usage* but still make the warning a bit more pointed?

Reply:
I've tried a rewrite to make this clearer, at the expense of being a lot more verbose :)